File: src/draft.ts
/**
 * N-gram drafter. Proposes whole chunks that a Noul can accept or reject.
 *
 * The measured reason this exists: one round trip costs the same whether you ask one
 * question or six, because questions inside a request run in parallel. So a free local
 * drafter plus one verification request per accepted chunk buys several tokens per
 * round trip instead of one.
 */

import { readFileSync } from "node:fs";

import type { Answers } from "./api.js";
import { END, Vocab } from "./vocab.js";

const PUNCTUATION = new Set([".", ",", "?", "!", ":", ";", "...", "-"]);

export class Drafter {
  readonly vocab: Vocab;
  readonly order: number;
  readonly maxLen: number;
  private readonly counts = new Map<string, Map<string, number>>();

  constructor(vocab: Vocab, order = 3, maxLen = 8) {
    this.vocab = vocab;
    this.order = order;
    this.maxLen = maxLen;
  }

  static fromText(text: string, vocab: Vocab, order = 3, maxLen = 8): Drafter {
    const drafter = new Drafter(vocab, order, maxLen);
    drafter.add(text);
    return drafter;
  }

  /** One document per line, so proposed chunks never cross a sentence break. */
  static fromFile(path: string, vocab: Vocab, order = 3, maxLen = 8): Drafter {
    const drafter = new Drafter(vocab, order, maxLen);
    for (const line of readFileSync(path, "utf8").split("\n")) {
      if (line.trim() !== "" && !line.trimStart().startsWith("#")) drafter.add(line);
    }
    return drafter;
  }

  add(text: string): void {
    const tokens = this.vocab.tokenize(text).filter((token): token is string => token !== null);
    for (let width = 1; width <= this.order; width += 1) {
      const keyWidth = width - 1;
      for (let index = 0; index + width <= tokens.length; index += 1) {
        const key = tokens.slice(index, index + keyWidth).join("\u0000");
        const next = tokens[index + keyWidth];
        if (next === undefined) continue;
        const counter = this.counts.get(key) ?? new Map<string, number>();
        counter.set(next, (counter.get(next) ?? 0) + 1);
        this.counts.set(key, counter);
      }
    }
  }

  /**
   * Longest-match backoff: try the full context, then drop the oldest token.
   *
   * `minWidth` stops the backoff before the empty context. Chunk extension uses
   * minWidth 1 so a chain cannot jump to an unrelated sentence.
   */
  nextToken(context: readonly string[], minWidth = 0): string | null {
    const widest = Math.min(context.length, this.order - 1);
    for (let width = widest; width >= minWidth; width -= 1) {
      const key = width === 0 ? "" : context.slice(context.length - width).join("\u0000");
      const counter = this.counts.get(key);
      if (counter !== undefined && counter.size > 0) return topToken(counter);
    }
    return null;
  }

  draw(context: readonly string[], limit = 2): string[] {
    const widest = Math.min(context.length, this.order - 1);
    for (let width = widest; width >= 0; width -= 1) {
      const key = width === 0 ? "" : context.slice(context.length - width).join("\u0000");
      const counter = this.counts.get(key);
      if (counter !== undefined && counter.size > 0) return topTokens(counter, limit);
    }
    return [];
  }

  /** Whole continuations to verify, longest first, up to k of them. */
  candidates(tokens: readonly string[], k = 6): string[] {
    const context = [...tokens];
    const chains: string[][] = [];
    for (const first of this.draw(context, 2)) {
      const chain = [first];
      chains.push([...chain]);
      for (let step = 0; step < this.maxLen - 1; step += 1) {
        const following = this.nextToken([...context, ...chain], 1);
        if (following === null || following === END) break;
        chain.push(following);
        chains.push([...chain]);
      }
    }

    const rendered: string[] = [];
    for (const chain of chains.sort((a, b) => b.length - a.length)) {
      const text = chain.join(" ");
      if (!rendered.includes(text)) rendered.push(text);
    }
    return rendered.slice(0, k);
  }
}

function topTokens(counter: ReadonlyMap<string, number>, limit: number): string[] {
  return [...counter.entries()]
    .sort(([, a], [, b]) => b - a)
    .slice(0, limit)
    .map(([token]) => token);
}

function topToken(counter: ReadonlyMap<string, number>): string | null {
  return topTokens(counter, 1)[0] ?? null;
}

export interface ChunkChoice {
  readonly chunk: string;
  readonly scored: readonly (readonly [string, number])[];
}

/**
 * Read the chunk Nouls out of one request and return the chunk to insert.
 *
 * Rule: keep the candidates scoring at least `keep` times the best score, above an
 * absolute `accept` floor, then take the longest survivor. Measured behavior:
 *
 * - Unrelated candidates separate well: `paris` 0.95 against `london` 0.02.
 * - Nested candidates barely separate, and the shortest wins slightly: `and` 0.44,
 *   `and looked` 0.42, `and looked at the street` 0.38. A length-agnostic rule picks one
 *   token; keep 0.8 picks the whole chain, which is the point of the drafter.
 * - Four invented continuations all landed at 0.14 to 0.17, below any sane floor.
 */
export function verifiedChunk(
  answers: Answers,
  candidates: readonly string[],
  accept = 0.25,
  keep = 0.8,
): ChunkChoice {
  const scored: [string, number][] = [];
  candidates.forEach((candidate, index) => {
    const answer = answers[`chunk_${index}`];
    if (answer !== undefined && answer.type === "noul") scored.push([candidate, answer.noul]);
  });
  if (scored.length === 0) return { chunk: "", scored };

  const top = Math.max(...scored.map(([, value]) => value));
  const threshold = Math.max(accept, top * keep);
  const qualified = scored.filter(([, value]) => value >= threshold);
  const ranked = [...scored].sort(([, a], [, b]) => b - a);
  if (qualified.length === 0) return { chunk: "", scored: ranked };

  let best = qualified[0];
  if (best === undefined) return { chunk: "", scored: ranked };
  for (const entry of qualified) {
    const [candidate, value] = entry;
    const [bestCandidate, bestValue] = best;
    if (
      countWords(candidate) > countWords(bestCandidate) ||
      (countWords(candidate) === countWords(bestCandidate) && value > bestValue)
    ) {
      best = entry;
    }
  }
  return { chunk: best[0], scored: ranked };
}

export function countWords(candidate: string): number {
  const words = candidate.split(" ").filter((word) => !PUNCTUATION.has(word));
  return (words.length > 0 ? words : candidate.split(" ")).length;
}

File: python/jev_cli/draft.py
"""N-gram drafter. Proposes whole chunks that a Noul can accept or reject.

The measured reason this exists: one round trip costs the same whether you ask one
question or six, because questions inside a request run in parallel. So a free local
drafter plus one verification request per accepted chunk buys several tokens per round
trip instead of one.
"""

from __future__ import annotations

from collections import Counter, defaultdict
from dataclasses import dataclass, field
from pathlib import Path

from .vocab import END, PUNCTUATION, Vocab


@dataclass
class Drafter:
    vocab: Vocab
    order: int = 3
    max_len: int = 8
    counts: dict[tuple[str, ...], Counter] = field(default_factory=lambda: defaultdict(Counter))

    @classmethod
    def from_text(cls, text: str, vocab: Vocab, order: int = 3, max_len: int = 8) -> "Drafter":
        drafter = cls(vocab=vocab, order=order, max_len=max_len)
        drafter.add(text)
        return drafter

    @classmethod
    def from_file(cls, path: str | Path, vocab: Vocab, order: int = 3, max_len: int = 8) -> "Drafter":
        """One document per line, so proposed chunks never cross a sentence break."""
        drafter = cls(vocab=vocab, order=order, max_len=max_len)
        for line in Path(path).read_text().splitlines():
            if line.strip() and not line.lstrip().startswith("#"):
                drafter.add(line)
        return drafter

    def add(self, text: str) -> None:
        tokens = [token for token in self.vocab.tokenize(text) if token]
        for width in range(1, self.order + 1):
            key_width = width - 1
            for index in range(len(tokens) - width + 1):
                key = tuple(tokens[index : index + key_width])
                self.counts[key][tokens[index + key_width]] += 1

    def next_token(self, context: tuple[str, ...], min_width: int = 0) -> str | None:
        """Longest-match backoff: try the full context, then drop the oldest token.

        min_width stops the backoff before it reaches the empty context. Chunk
        extension uses min_width=1 so a chain cannot jump to an unrelated sentence.
        """
        for width in range(min(len(context), self.order - 1), min_width - 1, -1):
            key = context[len(context) - width :] if width else ()
            counter = self.counts.get(key)
            if counter:
                return counter.most_common(1)[0][0]
        return None

    def draw(self, context: tuple[str, ...], limit: int = 2) -> list[str]:
        for width in range(min(len(context), self.order - 1), -1, -1):
            key = context[len(context) - width :] if width else ()
            counter = self.counts.get(key)
            if counter:
                return [token for token, _ in counter.most_common(limit)]
        return []

    def top_tokens(self, tokens: list[str], k: int = 40) -> list[str]:
        """The k most likely next tokens, by backoff from the longest context to the
        unigram table. This is the candidate set for a narrowed Choice question.

        Measured on held-out text: the true next token is inside the top 4 for 0.50 of
        positions, top 12 for 0.67, top 20 for 0.73, top 40 for 0.81, top 80 for 0.90.
        A narrow option map is a bet that pays off about four times in five at k=40.
        """
        out: list[str] = []
        for width in range(min(len(tokens), self.order - 1), -1, -1):
            key = tuple(tokens[len(tokens) - width :]) if width else ()
            counter = self.counts.get(key)
            if counter:
                for token, _ in counter.most_common(k):
                    if token not in out:
                        out.append(token)
            if len(out) >= k:
                break
        return out[:k]

    def candidates(self, tokens: list[str], k: int = 6) -> list[str]:
        """Whole continuations to verify, longest first, up to k of them."""
        context = tuple(tokens)
        chains: list[list[str]] = []
        for first in self.draw(context, limit=2):
            chain = [first]
            chains.append(list(chain))
            for _ in range(self.max_len - 1):
                following = self.next_token(context + tuple(chain), min_width=1)
                if following is None or following == END:
                    break
                chain.append(following)
                chains.append(list(chain))

        rendered: list[str] = []
        for chain in sorted(chains, key=len, reverse=True):
            text = " ".join(chain)
            if text not in rendered:
                rendered.append(text)
        return rendered[:k]


def verified_chunk(
    answers: dict,
    candidates: list[str],
    accept: float = 0.25,
    keep: float = 0.8,
) -> tuple[str, list[tuple[str, float]]]:
    """Read the chunk Nouls out of one request and return the chunk to insert, plus scores.

    Rule: keep the candidates scoring at least keep times the best score, above an
    absolute accept floor, then take the longest survivor. Measured behavior:

    * Unrelated candidates separate well: 'Paris' 0.95 against 'London' 0.02.
    * Nested candidates barely separate, and the shortest wins slightly: 'and' 0.44,
      'and looked' 0.42, 'and looked at the street' 0.38. A length-agnostic rule picks
      one token; keep=0.8 picks the whole chain, which is the point of the drafter.
    * Four invented continuations all landed at 0.14 to 0.17, below any sane floor.
    """
    scored = [
        (candidate, float(answers[f"chunk_{index}"]["noul"]))
        for index, candidate in enumerate(candidates)
        if answers.get(f"chunk_{index}", {}).get("type") == "noul"
    ]
    if not scored:
        return "", []
    top = max(value for _, value in scored)
    threshold = max(accept, top * keep)
    qualified = [(candidate, value) for candidate, value in scored if value >= threshold]
    if not qualified:
        return "", scored
    best = max(qualified, key=lambda pair: (count_words(pair[0]), pair[1]))[0]
    return best, sorted(scored, key=lambda pair: (-pair[1], -len(pair[0])))


def count_words(candidate: str) -> int:
    words = [word for word in candidate.split() if word not in PUNCTUATION]
    return len(words or candidate.split())
